Source/SpatialGDK/Private/SpatialInterop.cpp

```diff
@@ -860,16 +866,136 @@ bool USpatialInterop::IsSingletonClass(UClass* Class)
+
+	if (bCanSpawnReplicatedStablyNamedActors)
+	{
+		ReserveReplicatedStablyNamedActors();
+	}
+}
+
+void USpatialInterop::UnreserveReplicatedStablyNamedActor(AActor* Actor)
+{
+	// Slightly inefficient linear search, could use a hashmap
+	USpatialActorChannel* Channel = nullptr;
+	for (auto& It : ReplicatedStablyNamedActorChannelQueue)
+	{
+		if (It->Actor == Actor) {
+			Channel = It;
+			break;
+		}
+	}
+
+	if (Channel != nullptr) {
+		ReplicatedStablyNamedActorChannelQueue.Remove(Channel);
+	}
+}
+
+void USpatialInterop::AddReplicatedStablyNamedActorToGSM(const FEntityId& EntityId, AActor* Actor)
+{
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPathMap = *GetEntityIdToReplicatedStablyNamedPath();
+	// If the map already has the entity id, meaning the actor was already spawned, return early
+	if (EntityIdToReplicatedStablyNamedPathMap.count(EntityId.ToSpatialEntityId()) != 0) return;
+
+	std::string Path = TCHAR_TO_UTF8(*Actor->GetPathName(Actor->GetWorld()));
+	EntityIdToReplicatedStablyNamedPathMap.emplace(EntityId.ToSpatialEntityId(), Path);
+
+	improbable::unreal::GlobalStateManager::Update Update;
+	Update.set_entity_id_to_replicated_stably_named_path(EntityIdToReplicatedStablyNamedPathMap);
+
+	TSharedPtr<worker::Connection> Connection = SpatialOSInstance->GetConnection().Pin();
+	Connection->SendComponentUpdate<improbable::unreal::GlobalStateManager>(worker::EntityId((long)SpatialConstants::GLOBAL_STATE_MANAGER), Update);
+}
+
+void USpatialInterop::ReserveReplicatedStablyNamedActors()
+{
+	bCanSpawnReplicatedStablyNamedActors = true;
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPath = *GetEntityIdToReplicatedStablyNamedPath();
+
+	for (USpatialActorChannel* Channel : ReplicatedStablyNamedActorChannelQueue)
+	{
+		std::string Path = TCHAR_TO_UTF8(*Channel->Actor->GetPathName(Channel->Actor->GetWorld()));
+
+		// Inefficient, check if we have already spawned this actor
+		bool Found = false;
+		for (const auto& Pair : EntityIdToReplicatedStablyNamedPath)
+		{
+			if (Pair.second == Path)
+			{
+				Found = true;
+				Channel->RegisterEntityId(FEntityId{ Pair.first });
+				break;
+			}
+		}
+		if (Found)
+		{
+			continue;
+		}
+
+		Channel->SendReserveEntityIdRequest();
+	}
+
+	ReplicatedStablyNamedActorChannelQueue.Empty();
+}
+
+void USpatialInterop::DeleteIrrelevantReplicatedStablyNamedActors(const EntityIdToPathMap& EntityIdToReplicatedStablyNamedPath)
+{
+	for (auto& Pair : EntityIdToReplicatedStablyNamedPath)
+	{
+		FString FullPath = UTF8_TO_TCHAR(Pair.second.c_str());
+		AActor* Actor = FindObject<AActor>(GetWorld(), *FullPath);
+		worker::EntityId EntityId = Pair.first;
+
+		if (Actor != nullptr && ReplicatedStablyNamedActorTimeoutMap.Find(Actor) == nullptr)
+		{
+			FTimerHandle RetryTimer;
+			FTimerDelegate TimerCallback;
+			TimerCallback.BindLambda([Actor, EntityId, this] {
+				if (Actor != nullptr && Actor->IsFullNameStableForNetworking() && !PackageMap->GetNetGUIDFromEntityId(EntityId).IsValid())
```

Comment:
Actor's already been checked for null. And can't we check for stably named outside the lambda?

Reply:
Checking for Actor != null, in case it somehow gets deleted before the timeout gets to deleting it (not sure whether this is the correct flow, I have noticed Unreal somehow deletes the references to the Actor on deletion).
Agree on check, and this actually only gets run on stably named replicated actors (I hope), so the check is completely unnecessary.